File: src/leetcode/p0000_0099/p0069_sqrt/Sqrt.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package org.willxu.algorithm.service.integer;

public interface Sqrt {

    /**
     * Given a non-negative integer x, compute and return the square
     * root of x.
     *
     * Since the return type is an integer, the decimal digits are
     * truncated, and only te integer part of the result is returned.
     *
     * Note: You are not allowed to use any built-in exponent function
     * or operator, such as pow(x, 0.5) or x ** 0.5.
     *
     * @param x 0 <= x <= 2^31 - 1
     * @return square root of x.
     */
    int mySqrt(int x);
}

File: src/leetcode/p0000_0099/p0069_sqrt/sqrt.py
from abc import ABC, abstractmethod

class Sqrt(ABC):
    """
    Given a non-negative integer x, return the square root of x rounded down to the nearest integer. The returned integer should be non-negative as well.
    You must not use any built-in exponent function or operator.
    For example, do not use pow(x, 0.5) in c++ or x ** 0.5 in python.
    """
    @abstractmethod
    def my_sqrt(self, x: int) -> int:
        """
        Args:
            x:
                0 <= x <= 2^31 - 1
        """
        pass

class SqrtBinarySearch(Sqrt):
    def my_sqrt(self, x: int) -> int:
        if x < 2:
            return x
        left, right = 1, x // 2
        while left <= right:
            mid = (left + right) // 2
            if mid * mid == x:
                return mid
            elif mid * mid < x:
                left = mid + 1
            else:
                right = mid - 1
        return right
